nvidia_gpu_stats/cmd/nvidia_gpu_stats/main.go

```diff
@@ -0,0 +1,241 @@
+			key := fmt.Sprintf("gpu.%d.memoryTotal", di)
+			metrics[key] = memoryInfo.Total
+
+			key = fmt.Sprintf("gpu.%d.memoryAllocated", di)
+			metrics[key] = float64(memoryInfo.Used) / float64(memoryInfo.Total) * 100
+			if gpuInUseByProcess {
+				keyProc := fmt.Sprintf("gpu.process.%d.memoryAllocated", di)
+				metrics[keyProc] = metrics[key]
+			}
+
+			key = fmt.Sprintf("gpu.%d.memoryAllocatedBytes", di)
+			metrics[key] = float64(memoryInfo.Used)
+			if gpuInUseByProcess {
+				keyProc := fmt.Sprintf("gpu.process.%d.memoryAllocatedBytes", di)
+				metrics[keyProc] = metrics[key]
+			}
+		}
+
+		temperature, ret := device.GetTemperature(nvml.TEMPERATURE_GPU)
+		if ret == nvml.SUCCESS {
+			key := fmt.Sprintf("gpu.%d.temp", di)
+			metrics[key] = float64(temperature)
+			if gpuInUseByProcess {
+				keyProc := fmt.Sprintf("gpu.process.%d.temp", di)
+				metrics[keyProc] = metrics[key]
+			}
+		}
+
+		powerUsage, ret := device.GetPowerUsage()
+		if ret == nvml.SUCCESS {
+			key := fmt.Sprintf("gpu.%d.powerWatts", di)
+			metrics[key] = float64(powerUsage) / 1000
+			if gpuInUseByProcess {
+				keyProc := fmt.Sprintf("gpu.process.%d.powerWatts", di)
+				metrics[keyProc] = metrics[key]
+			}
+		}
+
+		powerLimit, ret := device.GetEnforcedPowerLimit()
+		if ret == nvml.SUCCESS {
+			key := fmt.Sprintf("gpu.%d.enforcedPowerLimitWatts", di)
+			metrics[key] = float64(powerLimit) / 1000
+			if gpuInUseByProcess {
+				keyProc := fmt.Sprintf("gpu.process.%d.enforcedPowerLimitWatts", di)
+				metrics[keyProc] = metrics[key]
+			}
+
+			key = fmt.Sprintf("gpu.%d.powerPercent", di)
+			metrics[key] = float64(powerUsage) / float64(powerLimit) * 100
+			if gpuInUseByProcess {
+				keyProc := fmt.Sprintf("gpu.process.%d.powerPercent", di)
+				metrics[keyProc] = metrics[key]
+			}
+		}
+	}
+
+	return metrics
+}
+
+func (g *GPUNvidia) IsAvailable() bool {
+	defer func() {
+		if r := recover(); r != nil {
+			g.nvmlInit = nvml.ERROR_UNINITIALIZED
+		}
+	}()
+	g.nvmlInit = nvml.Init()
+	return g.nvmlInit == nvml.SUCCESS
+}
+
+func (g *GPUNvidia) Close() {
+	err := nvml.Shutdown()
+	if err != nvml.SUCCESS {
+		return
+	}
+}
+
+func main() {
+	samplingInterval := flag.Duration("s", 1*time.Second, "sampling interval")
+	pid := flag.Int("pid", 0, "pid of the process to communicate with")
+
+	flag.Parse()
+
+	gpu := NewGPUNvidia(*pid)
+	defer gpu.Close()
+
+	if !gpu.IsAvailable() {
+		return
+	}
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	ticker := time.NewTicker(*samplingInterval)
+
+	done := make(chan struct{})
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				timeStamp := time.Now()
+				metrics := gpu.SampleMetrics()
+				if metrics == nil {
+					continue
+				}
+				metrics["_timestamp"] = float64(timeStamp.Unix()) + float64(timeStamp.Nanosecond())/1e9
+				output, err := json.Marshal(metrics)
```

Comment:
what do you think of having metrics struct (in core) with json hints, that way unmarshal on the other side would be into a struct and would make the code their simpler to handle?
just an idea